lib/vfscore: Fix create through broken symlink

andreittr opened the pull request:
### Description of changes

Previously a vfs syscall on a broken symlink would fail with ENOENT instead of attempting the syscall on the symlink target. This would make open(O_CREAT) to fail instead of creating the missing target file. This change fixes this behavior, bringing it in line with expectations.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

Test with:
```
symlink("nonexistent", "broken-link");
int f = open("broken-link", O_WRONLY|O_CREAT, 0644);
if (f < 0) perror("bug in open()");
```

andreittr committed:
lib/vfscore: Fix create through broken symlink

Previously a vfs syscall on a broken symlink would fail with ENOENT
instead of attempting the syscall on the symlink target. This would make
open(O_CREAT) to fail instead of creating the missing target file.
This change fixes this behavior, bringing it in line with expectations.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

RaduNichita commented on lib/vfscore/lookup.c:
```diff
@@ -223,21 +227,37 @@ namei(const char *path, struct dentry **dpp)
+ * @path: full path name.
+ * @dpp:  dentry to be returned.
+ */
+int
+namei(const char *path, struct dentry **dpp)
```
```suggestion
int namei(const char *path, struct dentry **dpp)
```

RaduNichita replied:
You can leave it on the same line, as mentioned in the coding - conventions PR: https://github.com/unikraft/docs/pull/248

> Prefer the Linux style with the return type on the same line as the function name.
In case of long functions + first argument, you can break after the return type.

andreittr replied:
All of lookup.c uses this format, and I believe maintaining consistency is important.
If we however believe this style actively harms readability, we can address it for the entire file in a style change PR.

In vfs.h: sure, there it makes sense, and the rest of the file follows suit as well.

mariasfiraiala commented on lib/vfscore/lookup.c:
```diff
@@ -96,13 +96,14 @@ namei_follow_link(struct dentry *dp, char *node, char *name, char *fp, size_t mo
  */
 int
-namei(const char *path, struct dentry **dpp)
+namei_resolve(const char *path, struct dentry **dpp, char realpath[PATH_MAX])
```
```suggestion
namei_resolve(const char *path, struct dentry **dpp, char *realpath)
```

Let's have this without the specified size, just like it happens for the other argument.

RaduNichita commented on lib/vfscore/vfs.h:
```diff
@@ -684,6 +684,22 @@ int sec_vnode_permission(char *path);
+ *	- (0):  Completed successfully
+ *	- (<0): Negative value with error code
+ */
+int
```
Same here

mariasfiraiala commented on lib/vfscore/lookup.c:
```diff
@@ -223,21 +227,37 @@ namei(const char *path, struct dentry **dpp)
-	return 0;
+out:
+	if (realpath)
+		strcpy(realpath, fp);
```
```suggestion
		strlcpy(realpath, fp, PATH_MAX);
```

We can use `strlcpy()` in order to keep the consistency [within the file](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/lookup.c#L87) and for better security.

RaduNichita commented on lib/vfscore/lookup.c:
```diff
@@ -223,21 +227,37 @@ namei(const char *path, struct dentry **dpp)
+	return error;
+}
+/*
+ * Convert a pathname into a pointer to a dentry
```
I think the commentary is not needed, since it's very similar in [vfs.h](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/vfs.h#L675)

andreittr replied:
Indeed the comments are duplicates of the docstrings in vfs.h, however the file already has this style for exported functions and I don't want to break it without good reason.

mariasfiraiala reviewed (changes requested):
Thanks @andreittr. See my few comments below. Other than that I wasn't able to reproduce the issue with the example you provided, though the changes make sense.

andreittr commented:
> Thanks @andreittr. See my few comments below. Other than that I wasn't able to reproduce the issue with the example you provided, though the changes make sense.

What behavior are you getting when trying to reproduce? If the symlink call fails you might need to apply this PR: https://github.com/unikraft/unikraft/pull/952.

mariasfiraiala commented:
> If the symlink call fails you might need to apply this PR: https://github.com/unikraft/unikraft/pull/952. 

I had already applied that PR since I also reviewed it.

> What behavior are you getting when trying to reproduce?

The bug is not reproducible on `9pfs` (meaning that the error message doesn't get printed), but with `ramfs`, the behaviour is indeed the one you mentioned, so all good.

@andreittr I'll approve the PR after you apply the suggestions, thank you for this!

andreittr commented:
Pushed changes addressing comments.

> The bug is not reproducible on `9pfs` (meaning that the error message doesn't get printed), but with `ramfs`, the behaviour is indeed the one you mentioned, so all good.

That's really odd that 9pfs isn't affected, as the code paths are common between filesystems :shrug: .

mariasfiraiala reviewed (approved):
All good on my side now, thanks a lot for the fixes @andreittr!

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

RaduNichita commented:
All good also from my side. As @mariasfiraiala, I have only managed to reproduce the bug on ramfs, not 9pfs.

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
